nixos/modules/services/admin/pgadmin4.nix

```diff
@@ -0,0 +1,106 @@
+with lib;
+
+let
+  pkg = pkgs.pgadmin4;
+  cfg = config.services.pgadmin4;
+
+  base = with types; oneOf [ int bool str ];
+
+  formatAttrset = attr:
+    "{${concatStringsSep "\n" (mapAttrsToList (key: value: "${builtins.toJSON key}: ${formatPyValue value},") attr)}}";
+
+  formatPyValue = value:
+    if builtins.isString value then builtins.toJSON value
+    else if value ? _expr then value._expr
+    else if builtins.isInt value then toString value
+    else if builtins.isBool value then (if value then "True" else "False")
+    else if builtins.isAttrset value then (formatAttrset value)
+    else if builtins.isList value then "[${concatStringsSep "\n" (map formatPyValue value)}]"
+    else throw "Unrecognized type";
+
+  formatPy = attrs:
+    concatStringsSep "\n" (mapAttrsToList (key: value: "${key} = ${formatPyValue value}") attrs);
+
+  pyType = with types; attrsOf (oneOf [ (listOf base) (attrsOf base) base ]);
+in
+{
+  options.services.pgadmin4 = {
+    enable = mkEnableOption "PostgreSQL Admin 4";
+
+    port = mkOption {
+      description = "Port for pgadmin4 to run on";
+      type = types.port;
+      default = 5050;
+    };
+
+    initialEmail = mkOption {
+      description = "Initial email for the pgAdmin account.";
+      type = types.str;
+    };
+
+    initialPassword = mkOption {
+      description = "Initial password for the pgAdmin account.";
+      type = types.str;
+    };
+
+    openFirewall = mkEnableOption "firewall passthrough for pgadmin4";
+
+    settings = mkOption {
+      description = ''
+        Settings for pgadmin4.
+        <link xlink:href="https://www.pgadmin.org/docs/pgadmin4/development/config_py.html">Documentation</link>.
+      '';
+      type = pyType;
+      default= {};
+    };
+  };
+
+  config = mkIf (cfg.enable) {
+    networking.firewall.allowedTCPPorts = mkIf (cfg.openFirewall) [ cfg.port ];
+
+    services.pgadmin4.settings = {
+      DEFAULT_SERVER_PORT = cfg.port;
+      SERVER_MODE = true;
+    } // (optionalAttrs cfg.openFirewall {
+      DEFAULT_SERVER = mkDefault "::";
+    });
+
+    services.postgresql = {
+
+    };
+
+    systemd.services.pgadmin4 = {
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+      # requires = [ "network.target" "postgresql.service" ];
+      requires = [ "network.target" ];
+      wants = [ "postgresql.service" ]; # might not exist
+
+      path = [ config.services.postgresql.package pkgs.coreutils pkgs.bash ];
+
+      preStart = ''
+        # TODO: rewrite to expect script?
+        (
+          # Email address:
+          echo ${escapeShellArg cfg.initialEmail}
+          # Password:
+          echo ${escapeShellArg cfg.initialPassword}
+          # Retype password:
+          echo ${escapeShellArg cfg.initialPassword}
+        ) | ${pkg}/bin/pgadmin4-setup
+      '';
+
+      serviceConfig = {
+        User = "pgadmin4";
+        DynamicUser = true;
+        LogsDirectory = "pgadmin";
+        StateDirectory = "pgadmin";
```

Comment:
shouldn't we use the values from the config?

Reply:
This here is so systemd auto-creates the directories. There's no way to specify a full path here, only the suffix. The user would be responsible for creating the custom directories.

Reply:
I'm not really familiar with serviceConfig. Where does StateDirectory and LogsDirectory get used?
I'd kinda prefer a (known) location such as /var/lib/pgadmin or a location specified by configuration.nix

Reply:
Ah ok, that sounds exactly like what we want :+1: